### PR TITLE
Implement file sharing

### PR DESCRIPTION
**Pull request by eliasnaur**

### CODE REVIEW

**Comment by bradfitz** [comment]
Yay! I'll take a look in a few hours.

**Review by DentonGentry — COMMENTED** [review]
I downloaded this, built an APK, and have been able to share files back and forth with a Macbook. I've sent images and PDF files, I've sent one file and four files at a time. Everything has worked well, files arrive quickly and notifications appear when files are received.

File: android/src/main/java/com/tailscale/ipn/IPNActivity.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
```
Comment (Xe) [inline android/src/main/java/com/tailscale/ipn/IPNActivity.java:1]:
```suggestion
// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
```

File: jni/jni.go [hunk 2]
```diff
@@ -343,6 +462,38 @@ func DeleteGlobalRef(e *Env, obj Object) {
...
+	cstr := C.CString(name)
+	defer C.free(unsafe.Pointer(cstr))
+	cls := C.jni_FindClass(env(e), cstr)
+	if cls == 0 {
+		panic(fmt.Errorf("Java class not found: %s", name))
+	}
+	return Class(cls)
+}
+
+// GoValue converts obj to a Go value. It returns nil for unsupported types.
```
Comment (bradfitz) [inline jni/jni.go:475]: list which are supported?

File: jni/jni.go [hunk 3]
```diff
@@ -343,6 +462,38 @@ func DeleteGlobalRef(e *Env, obj Object) {
 	C.jni_DeleteGlobalRef(env(e), C.jobject(obj))
 }
 
+func findClass(e *Env, name string) Class {
```
Comment (bradfitz) [inline jni/jni.go:465]: give an example input string for `name`? e.g. `"java/lang/String"` (not period delimited, no `L` prefix, ...)

File: jni/jni.go [hunk 4]
```diff
@@ -236,6 +285,76 @@ func GetByteArrayElements(e *Env, jarr ByteArray) []byte {
...
 
+// GetBooleanArrayElements returns the contents of the boolean array.
+func GetBooleanArrayElements(e *Env, jarr BooleanArray) []bool {
+	if jarr == 0 {
+		return nil
+	}
+	size := C.jni_GetArrayLength(env(e), C.jarray(jarr))
+	elems := C.jni_GetBooleanArrayElements(env(e), C.jbooleanArray(jarr))
+	defer C.jni_ReleaseBooleanArrayElements(env(e), C.jbooleanArray(jarr), elems, 0)
+	backing := (*(*[1 << 30]C.jboolean)(unsafe.Pointer(elems)))[:size:size]
```
Comment (bradfitz) [inline jni/jni.go:296]: I can't wait for unsafe.Slice, etc: https://tip.golang.org/doc/go1.17#language

Hopefully we can convert these soon enough.

File: cmd/tailscale/main.go [hunk 5]
```diff
@@ -114,13 +167,20 @@ type OAuth2Event struct {
 	Token *tailcfg.Oauth2Token
 }
 
+type FileSendEvent struct {
+	Target  *apitype.FileTarget
+	Context context.Context
+	Updates func(info FileSendInfo)
```
Comment (bradfitz) [inline cmd/tailscale/main.go:173]: `info` can be dropped... doesn't add anything considering the type name

File: cmd/tailscale/main.go [hunk 6]
```diff
@@ -385,6 +481,89 @@ func (a *App) runBackend() error {
...
+	var mediaURI string
+	err = jni.Do(a.jvm, func(env *jni.Env) error {
+		cls := jni.GetObjectClass(env, a.appCtx)
+		insertMedia := jni.GetMethodID(env, cls, "insertMedia", "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;")
+		jname := jni.JavaString(env, f.Name)
+		jmime := jni.JavaString(env, mimeType)
+		uri, err := jni.CallObjectMethod(env, a.appCtx, insertMedia, jni.Value(jname), jni.Value(jmime))
+		if err != nil {
+			return err
+
```
Comment (bradfitz) [inline cmd/tailscale/main.go:515]: leftover blank line

File: cmd/tailscale/main.go [hunk 7]
```diff
@@ -385,6 +481,89 @@ func (a *App) runBackend() error {
...
+		deleteURI(mediaURI)
+		return fmt.Errorf("close: %w", err)
+	}
+	if err := a.notifyFile(mediaURI, f.Name); err != nil {
+		fatalErr(err)
+	}
+	return b.DeleteFile(f.Name)
+}
+
+func (a *App) openURI(uri, mode string) (*os.File, error) {
```
Comment (bradfitz) [inline cmd/tailscale/main.go:550]: add a comment that references the Android API this wraps, so readers know what "mode" can be? Oh, this is just calling into its own Java, not the SDK.

k, add a comment about mode then?

File: cmd/tailscale/main.go [hunk 8]
```diff
@@ -858,8 +1059,108 @@ func (a *App) processUIEvents(w *app.Window, events []UIEvent, act jni.Object, s
...
 			state.query = strings.ToLower(e.Query)
 			a.updateState(act, state)
+		case FileSendEvent:
+			a.sendFiles(e, files)
+		}
+	}
+}
+
+func (a *App) sendFiles(e FileSendEvent, files []File) {
+	go func() {
```
Comment (bradfitz) [inline cmd/tailscale/main.go:1069]: I'd prefer longish-lived goroutines to have names. Makes reading debug stacks easier.

`go a.sendFilesAsync(...)` ? or `Routine`, etc.

File: cmd/tailscale/main.go [hunk 9]
```diff
@@ -858,8 +1059,108 @@ func (a *App) processUIEvents(w *app.Window, events []UIEvent, act jni.Object, s
...
+			})
+			a.invalidate()
+		}
+		defer a.invalidate()
+		for _, f := range files {
+			if err := a.sendFile(e.Context, e.Target, f, progress); err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+				log.Printf("file send failed: %v", err)
```
Comment (bradfitz) [inline cmd/tailscale/main.go:1092]: be careful about logging filenames. we scrub filenames (and even exact sizes) from all log output in the clients, so we can't know what's being shared.

File: cmd/tailscale/main.go [hunk 10]
```diff
@@ -858,8 +1059,108 @@ func (a *App) processUIEvents(w *app.Window, events []UIEvent, act jni.Object, s
...
+		body = strings.NewReader(f.Text)
+	case FileTypeURI:
+		f, err := a.openURI(f.URI, "r")
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		body = f
+	default:
+		panic("unknown file type")
```
Comment (bradfitz) [inline cmd/tailscale/main.go:1125]: panic? does returning an error work instead?

File: cmd/tailscale/main.go [hunk 11]
```diff
@@ -858,8 +1059,108 @@ func (a *App) processUIEvents(w *app.Window, events []UIEvent, act jni.Object, s
...
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
+		return fmt.Errorf("PUT failed: %s", res.Status)
+	}
+	return nil
+}
+
+type progressReader struct {
```
Comment (bradfitz) [inline cmd/tailscale/main.go:1145]: add some docs on this type

File: jni/jni.go (hunk 2 above)
Comment (eliasnaur) [reply jni/jni.go:475]: Deleted. I didn't end up using it.

File: cmd/tailscale/main.go (hunk 9 above)
Comment (eliasnaur) [reply cmd/tailscale/main.go:1092]: Removed logging. The UI will show the error anyway.

File: cmd/tailscale/main.go (hunk 10 above)
Comment (eliasnaur) [reply cmd/tailscale/main.go:1125]: The default branch indicates a programming error; we're in control of the possible enum values.

File: jni/jni.go (hunk 3 above)
Comment (eliasnaur) [reply jni/jni.go:465]: Deleted.

File: cmd/tailscale/main.go (hunk 8 above)
Comment (eliasnaur) [reply cmd/tailscale/main.go:1069]: Routine is good. Thanks.

**Comment by eliasnaur** [comment]
Ok to merge? Shall I do another release or do you want to?

**Comment by bradfitz** [comment]
Yeah, LGTM. And yes, a release would be great, thanks!

**Comment by bradfitz** [comment]
I get an error building this:

```
root@63968e99d1a6:~/tailscale-android# make tailscale-debug.apk
mkdir -p android/libs
go run gioui.org/cmd/gogio -buildmode archive -target android -appid com.tailscale.ipn -o android/libs/ipn.aar github.com/tailscale/tailscale-android/cmd/tailscale
go: downloading gioui.org v0.0.0-20210626160534-a87206c3647c
go: downloading gioui.org/cmd v0.0.0-20210623091900-5b8da35a798b
go: downloading github.com/akavel/rsrc v0.10.1
gogio: /usr/local/openjdk-8/bin/javac -target 1.8 -source 1.8 -sourcepath /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm -bootclasspath /build/android-sdk/platforms/android-29/android.jar -d /tmp/gogio-469261090/classes /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/Gio.java /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioActivity.java /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java failed: /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:33: error: cannot find symbol
import android.view.WindowInsetsController;
                   ^
  symbol:   class WindowInsetsController
  location: package android.view
/build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:184: error: cannot find symbol
            insetsMask = WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS;
                         ^
  symbol:   variable WindowInsetsController
  location: class GioView
/build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:189: error: cannot find symbol
            insetsMask = WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS;
                         ^
  symbol:   variable WindowInsetsController
  location: class GioView
/build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:201: error: cannot find symbol
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
                                                       ^
  symbol:   variable R
  location: class VERSION_CODES
/build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:212: error: cannot find symbol
        WindowInsetsController insetsController = window.getInsetsController();
        ^
  symbol:   class WindowInsetsController
  location: class GioView
/build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java:212: error: cannot find symbol
        WindowInsetsController insetsController = window.getInsetsController();
                                                        ^
  symbol:   method getInsetsController()
  location: variable window of type Window
Note: /build/go/pkg/mod/gioui.org@v0.0.0-20210626160534-a87206c3647c/app/internal/wm/GioView.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
6 errors

exit status 1
make: *** [Makefile:35: tailscale-debug.apk] Error 1
```

**Comment by bradfitz** [comment]
WindowInsetsController was added in API level 30.

And we have:

```
android {
    compileSdkVersion 29
    defaultConfig {
        minSdkVersion 22
        targetSdkVersion 29
```

**Comment by bradfitz** [comment]
@rosszurowski wants to play with an APK before we ship it to users, but I can't build the APK for him due to the dockershell `make tailscale-debug.apk` being broken right now.

**Comment by bradfitz** [comment]
Still doesn't work. Same error.

**Comment by eliasnaur** [comment]
Try again, I adjusted the Dockerfile as well.

**Comment by bradfitz** [comment]
That works. I sent @rosszurowski an APK.

**Comment by bradfitz** [comment]
Okay, all good here. Release at will. Thanks!

**Comment by eliasnaur** [comment]
Done.

**Pull request merged**